src/core/layout/qgslayoutitemattributetable.cpp

```diff
@@ -518,6 +518,12 @@ bool QgsLayoutItemAttributeTable::getTableContents( QgsLayoutTableContents &cont
     QgsFeature atlasFeature = mLayout->reportContext().feature();
     req.setFilterFid( atlasFeature.id() );
   }
+  
+  QVector< QPair<int, bool> > sortColumns = sortAttributes();
+  for ( int i = sortColumns.size() - 1; i >= 0; --i )
+  {
+     req = req.addOrderBy( mColumns.at( sortColumns.at( i ).first )->heading(), sortColumns.at( i ).second );
```

Comment:
`heading` isn't appropriate here -- that's a free text field, unrelated to the column's content

Reply:
Ah right, I see what you mean, I'll get this sorted out tomorrow with the indentation and the commit comment. Thanks!

Reply:
Thanks! and yes, a test would be valuable for this too (better do a range of tests -- one with a attribute only column, with with expression based column, and one with descending sort order)

Reply:
I'm looking at the code and they way to set sort columns in QgsLayoutItemAttributeTable seem to be done in the GUI side of things as there are no ways to set this in the base class. The columns to handle sorting is in qgslayoutattributeselectiondialog.cpp I would think that implementing functional test would require a revisit of the current handling or a new suite of test relying on gui elements. I think that this is out of the scope of a simple bugfix. Let me know what you think.

Reply:
Maybe I'm bad at this but classes are so convoluted I can't even find a way to make an new column with an expression with the exposed api. Sorting might be doable but I can't access the columns to perform sorting, again an api rework would be needed. Currently my changes work with minimal impact. I think we'll have to leave it at that until tests are easier to perform. :/

Reply:
It's not that hard. No API work necessary full stop. Look at TestQgsLayoutTable::attributeTableSetAttributes() and do something like

    table->columns().at( 0 )->setAttribute( "some expression which messes with a sort order like taking 0 - an int column value" );

Reply:
Ah right, I was expecting a function to expose things and create a new column, instead of replacing a column but this will do.

Reply:
Tests are done